client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java

```diff
@@ -500,7 +501,7 @@ private void reconnect(
             if (connectionFailureCause == null) {
                 connectionHandler.handle(Future.failedFuture(
                         new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "failed to connect")));
-            } else if (connectionFailureCause instanceof SecurityException) {
+            } else if (connectionFailureCause instanceof SaslSystemException) {
```

Comment:
it is true that the behavior in vertx-proton has changed. However, `ProtonSaslClientAuthenticatorImpl` uses different exceptions to indicate different types of failures:
```
  private void handleSaslFail() throws SaslException {
    switch(sasl.getOutcome()) {
    case PN_SASL_AUTH:
      throw new AuthenticationException("Failed to authenticate");
    case PN_SASL_SYS:
    case PN_SASL_TEMP:
      throw new SaslSystemException(false, "SASL handshake failed due to a transient error");
    case PN_SASL_PERM:
      throw new SaslSystemException(true, "SASL handshake failed due to an unrecoverable error");
    default:
      throw new SaslException("SASL handshake failed");
    }
  }
```
In particular, the most probable cause for failure (wrong credentials) is indicated by a `javax.security.sasl.AuthenticationException` (a subclass of `javax.security.sasl.SaslException`).

So, I think that we should catch `javax.security.sasl.SaslException` here instead of `SecurityException` and add another catch clause for `SaslSystemException` which is used to indicate cases where the server is not ready to process the handshake and fail with a `ServerErrorException` in that case.

Reply:
right, I now see I was concentrating on only one case, the one thrown in `handleSaslInit` if no SASL mechanism is found: 
````
throw new SaslSystemException(
            true, "Could not find a suitable SASL mechanism for the remote peer using the available credentials.");
````

A further note about this:
That case can also occur if the PLAIN mechanism is used and no username/password was given. 
In general, I would prefer to throw a `ClientErrorException` in that case - also keeping the behaviour from before the vert.x update. WDYT?

(Unfortunately, to distinguish that case from the `SASL handshake failed due to an unrecoverable error` case, a match on the exception text is required. We could add a small integration test to be safe concerning future vert.x changes here.)

Reply:
Giving it some more thought, I think that the `reconnect()` method in general should
* fail the attempt with a `ClientErrorException` (e.g. 401) in case of an `AuthenticationException`
* retry to connect in case of a `SaslSystemException` with *permanent* being `false`
* fail the attempt with a `ServerErrorException` (e.g. 503) in case of a `SaslSystemException` with *permanent* being `true`
* retry to connect in case of another `SaslException`

Regarding the case where no SASL mechanism is found, I agree that this should cause a `ClientErrorException` and that we would need to match on the exception's description (for the time being). I think that proton-vertx should throw an `AuthenticationException` in this case instead of a `SaslSystemException`. I will create a PR for that. Thanks for spotting this :+1:

Reply:
I agree that it's better to throw `AuthenticationException` in `handleSaslInit`.

Regarding your list of cases above:
By "fail the attempt", do you mean fail the overall connection attempt, ignoring the `reconnectAttempts`, and doing no further connection attempts?

In the "retry to connect" cases, do you mean reconnect without taking the `reconnectAttempts` into account, potentially doing reconnect attempts indefinitely?

Reply:
I've updated the commit - now failing directly on `AuthenticationException` and permanent `SaslSystemException`.